opnode/node/api.go

```diff
@@ -74,3 +90,105 @@ func toBlockNumArg(number rpc.BlockNumber) string {
+	PrevL2BlockHash common.Hash
+	PrevL2BlockNum  hexutil.Uint64
+
+	// LastL2BlockHash is the L2 block hash of the last block in the bundle.
+	// This is the ideal continuation point for the next batch submission.
+	// It will equal PrevL2BlockHash if there are no batches to submit.
+	LastL2BlockHash common.Hash
+	LastL2BlockNum  hexutil.Uint64
+
+	// Bundle represents the encoded bundle of batches.
+	// Each batch represents the inputs of a L2 block, i.e. a batch of L2 transactions (excl. deposits and such).
+	// The bundle encoding supports versioning and compression.
+	// The rollup-node determines the version to use based on configuration.
+	// Bundle is empty if there is nothing to submit.
+	Bundle hexutil.Bytes
+}
+
+func (n *nodeAPI) GetBatchBundle(ctx context.Context, req *BatchBundleRequest) (*BatchBundleResponse, error) {
+	var found eth.BlockID
+	// First find the common point with L2 history so far
+	for i, h := range req.L2History {
+		l2Ref, err := n.client.L2BlockRefByHash(ctx, h)
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) { // on reorgs and such we expect that blocks may be missing
+				continue
+			}
+			return nil, fmt.Errorf("failed to check L2 history for block hash %d in request %s: %v", i, h, err)
+		}
+		// found a block that exists! Now make sure it's really a canonical block of L2
+		canonBlock, err := n.client.L2BlockRefByNumber(ctx, big.NewInt(int64(l2Ref.Number)))
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) {
+				continue
+			}
+			return nil, fmt.Errorf("failed to check L2 history for block number %d, expecting block %s: %v", l2Ref.Number, h, err)
+		}
+		if canonBlock.Hash == h {
+			// found a common canonical block!
+			found = eth.BlockID{Hash: canonBlock.Hash, Number: canonBlock.Number}
+			break
+		}
+	}
+	if found == (eth.BlockID{}) { // none of the L2 history could be found. At least
+		return nil, ethereum.NotFound
+	}
+
+	lastL2Hash := found.Hash
+	lastL2BlockNum := found.Number
+	var batches []*derive.BatchData
+	// Now continue fetching the next blocks, and build batches, until we either run out of space, or run out of blocks.
+	for i := found.Number + 1; i < found.Number+MaxL2BlocksPerBatchResponse+1; i++ {
+		l2Block, err := n.client.BlockByNumber(ctx, big.NewInt(int64(i)))
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) { // block number too high
+				break
+			}
+			return nil, fmt.Errorf("failed to retrieve L2 block by number %d: %v", i, err)
+		}
+		batch, err := l2.BlockToBatch(n.config, l2Block)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert L2 block %d (%s) to batch: %v", i, l2Block.Hash(), err)
+		}
+		lastL2Hash = l2Block.Hash()
+		lastL2BlockNum = l2Block.Number().Uint64()
+		if batch == nil { // empty block, nothing to submit as batch
+			continue
+		}
+		batches = append(batches, batch)
+		// TODO: estimate size of all batches so far (including compressing them together, if we reached max size yet)
```

Comment:
TODO: we're should not append the batch, and break this loop, once the total size gets too large.

Reply:
Currently I think the only accurate way to do this is to encode the full set with each loop iteration. Is there a better way to estimate the size of each incremental batch into a running total? Do you think its sufficient to just add up the length of `rlp.Encode(batch)`?

Reply:
The current thing looks ok to me. And this is out of protocol consensus anyway. If we want we can optimize this later.